tests/test_bounce_integral.py

```diff
@@ -0,0 +1,828 @@
+        np.testing.assert_allclose(bp1, intersect[[0, -2]])
+        np.testing.assert_allclose(bp2, intersect[[1, -1]])
+
+    def test_extrema_first_and_before_bp1(plot=False):
+        start = -1.2 * np.pi
+        end = -2 * start
+        k = np.linspace(start, end, 7)
+        B = CubicHermiteSpline(
+            k,
+            np.cos(k) + 2 * np.sin(-2 * k) + k / 20,
+            -np.sin(k) - 4 * np.cos(-2 * k) + 1 / 20,
+        )
+        B_z_ra = B.derivative()
+        pitch = 1 / B(B_z_ra.roots(extrapolate=False))[2]
+        if plot:
+            plot_field_line(B, pitch, k[2], end)
+
+        bp1, bp2 = bounce_points(pitch, k[2:], B.c[:, 2:], B_z_ra.c[:, 2:], check=True)
+        bp1, bp2 = map(_filter_not_nan, (bp1, bp2))
+        # Our routine correctly detects intersection, while scipy, jnp.root fails.
+        intersect = B.solve(1 / pitch, extrapolate=False)
+        np.testing.assert_allclose(bp1[0], 0.8353192766102349)
+        intersect = np.insert(intersect, np.searchsorted(intersect, bp1[0]), bp1[0])
+        intersect = intersect[intersect >= k[2]]
+        np.testing.assert_allclose(bp1, intersect[[0, 1, 3]])
+        np.testing.assert_allclose(bp2, intersect[[0, 2, 4]])
+
+    def test_extrema_first_and_before_bp2(plot=False):
+        start = -1.2 * np.pi
+        end = -2 * start + 1
+        k = np.linspace(start, end, 7)
+        B = CubicHermiteSpline(
+            k,
+            np.cos(k) + 2 * np.sin(-2 * k) + k / 10,
+            -np.sin(k) - 4 * np.cos(-2 * k) + 1 / 10,
+        )
+        B_z_ra = B.derivative()
+        pitch = 1 / B(B_z_ra.roots(extrapolate=False))[1]
+        if plot:
+            plot_field_line(B, pitch, start, end)
+
+        bp1, bp2 = bounce_points(pitch, k, B.c, B_z_ra.c, check=True)
+        bp1, bp2 = map(_filter_not_nan, (bp1, bp2))
+        # Our routine correctly detects intersection, while scipy, jnp.root fails.
+        intersect = B.solve(1 / pitch, extrapolate=False)
+        np.testing.assert_allclose(bp1[0], -0.6719044147510538)
+        intersect = np.insert(intersect, np.searchsorted(intersect, bp1[0]), bp1[0])
+        np.testing.assert_allclose(bp1, intersect[0::2])
+        np.testing.assert_allclose(bp2, intersect[1::2])
+
+    # These are all the unique cases, if all tests pass then the bounce_points
+    # should work correctly for all inputs. Pass in True to see plots.
+    test_bp1_first()
+    test_bp2_first()
+    test_bp1_before_extrema()
+    test_bp2_before_extrema()
+    # In theory, this test should only pass if distinct=True when computing the
+    # intersections in bounce points. However, we can get lucky due to floating
+    # point errors, and it may also pass when distinct=False.
+    test_extrema_first_and_before_bp1()
+    test_extrema_first_and_before_bp2()
+
+
+@pytest.mark.unit
+def test_automorphism():
+    """Test automorphisms."""
+    a, b = -312, 786
+    x = np.linspace(a, b, 10)
+    y = _affine_bijection_forward(x, a, b)
```

Comment:
I think the function name doesn't have an _ in the front.

This happens with some other functions in this test as well.

Reply:
I had renamed these functions, so it's possible your local copy is out of date.

Reply:
Ah, ok. I thought I pulled the latest copy. Maybe not!